File: FCL/src/main/java/com/tungsten/fcl/ui/manage/LocalModListAdapter.java
package com.tungsten.fcl.ui.manage;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.BitmapFactory;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.Glide;
import com.tungsten.fcl.R;
import com.tungsten.fclcore.fakefx.beans.InvalidationListener;
import com.tungsten.fclcore.fakefx.beans.property.BooleanProperty;
import com.tungsten.fclcore.fakefx.beans.property.ListProperty;
import com.tungsten.fclcore.fakefx.beans.property.SimpleListProperty;
import com.tungsten.fclcore.fakefx.collections.FXCollections;
import com.tungsten.fclcore.mod.ModLoaderType;
import com.tungsten.fclcore.mod.RemoteMod;
import com.tungsten.fclcore.task.Schedulers;
import com.tungsten.fclcore.task.Task;
import com.tungsten.fclcore.util.StringUtils;
import com.tungsten.fclcore.util.io.CompressingUtils;
import com.tungsten.fcllibrary.component.FCLAdapter;
import com.tungsten.fcllibrary.component.theme.ThemeEngine;
import com.tungsten.fcllibrary.component.view.FCLCheckBox;
import com.tungsten.fcllibrary.component.view.FCLImageButton;
import com.tungsten.fcllibrary.component.view.FCLImageView;
import com.tungsten.fcllibrary.component.view.FCLLinearLayout;
import com.tungsten.fcllibrary.component.view.FCLTextView;
import com.tungsten.fcllibrary.util.LocaleUtils;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Optional;

public class LocalModListAdapter extends FCLAdapter {

    private final ModListPage modListPage;

    private final ListProperty<ModListPage.ModInfoObject> listProperty = new SimpleListProperty<>(FXCollections.observableArrayList());
    private final ListProperty<ModListPage.ModInfoObject> selectedItemsProperty = new SimpleListProperty<>(FXCollections.observableArrayList());

    public ListProperty<ModListPage.ModInfoObject> listProperty() {
        return listProperty;
    }

    public ListProperty<ModListPage.ModInfoObject> selectedItemsProperty() {
        return selectedItemsProperty;
    }

    public void selectAll() {
        selectedItemsProperty.clear();
        selectedItemsProperty.addAll(listProperty);
    }

    private boolean fromSelf = false;

    public LocalModListAdapter(Context context, ModListPage modListPage) {
        super(context);
        this.modListPage = modListPage;

        this.listProperty.addListener((InvalidationListener) observable -> {
            fromSelf = true;
            selectedItemsProperty.clear();
            fromSelf = false;
            notifyDataSetChanged();
        });
        selectedItemsProperty.addListener((InvalidationListener) observable -> {
            if (!fromSelf) {
                notifyDataSetChanged();
            }
        });
    }

    private static class ViewHolder {
        FCLLinearLayout parent;
        FCLCheckBox checkBox;
        FCLImageView icon;
        FCLTextView name;
        FCLTextView tag;
        FCLTextView description;
        FCLImageButton restore;
        FCLImageButton info;
        BooleanProperty booleanProperty;
    }

    @Override
    public int getCount() {
        return listProperty.getSize();
    }

    @Override
    public Object getItem(int i) {
        return listProperty.get(i);
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        final ViewHolder viewHolder;
        if (view == null) {
            viewHolder = new ViewHolder();
            view = LayoutInflater.from(getContext()).inflate(R.layout.item_local_mod, null);
            viewHolder.parent = view.findViewById(R.id.parent);
            viewHolder.checkBox = view.findViewById(R.id.check);
            viewHolder.icon = view.findViewById(R.id.icon);
            viewHolder.name = view.findViewById(R.id.name);
            viewHolder.tag = view.findViewById(R.id.tag);
            viewHolder.description = view.findViewById(R.id.description);
            viewHolder.restore = view.findViewById(R.id.restore);
            viewHolder.info = view.findViewById(R.id.info);
            view.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) view.getTag();
        }
        ModListPage.ModInfoObject modInfoObject = listProperty.get(i);
        viewHolder.parent.setBackgroundTintList(new ColorStateList(new int[][]{{}}, new int[]{selectedItemsProperty.contains(modInfoObject) ? ThemeEngine.getInstance().getTheme().getColor() : ThemeEngine.getInstance().getTheme().getLtColor()}));
        ThemeEngine.getInstance().registerEvent(viewHolder.parent, () -> viewHolder.parent.setBackgroundTintList(new ColorStateList(new int[][]{{}}, new int[]{selectedItemsProperty.contains(modInfoObject) ? ThemeEngine.getInstance().getTheme().getColor() : ThemeEngine.getInstance().getTheme().getLtColor()})));
        viewHolder.parent.setOnClickListener(v -> {
            if (selectedItemsProperty.contains(modInfoObject)) {
                fromSelf = true;
                selectedItemsProperty.remove(modInfoObject);
                fromSelf = false;
                viewHolder.parent.setBackgroundTintList(new ColorStateList(new int[][]{{}}, new int[]{ThemeEngine.getInstance().getTheme().getLtColor()}));
            } else {
                fromSelf = true;
                selectedItemsProperty.add(modInfoObject);
                fromSelf = false;
                viewHolder.parent.setBackgroundTintList(new ColorStateList(new int[][]{{}}, new int[]{ThemeEngine.getInstance().getTheme().getColor()}));
            }
        });
        viewHolder.checkBox.addCheckedChangeListener();
        if (viewHolder.booleanProperty != null) {
            viewHolder.checkBox.checkProperty().unbindBidirectional(viewHolder.booleanProperty);
        }
        viewHolder.checkBox.checkProperty().bindBidirectional(viewHolder.booleanProperty = modInfoObject.getActive());
        viewHolder.icon.setTag(i);
        viewHolder.icon.setImageBitmap(null);
        viewHolder.icon.setVisibility(View.GONE);
        viewHolder.name.setText(modInfoObject.getTitle());
        String tag = getTag(modInfoObject);
        viewHolder.tag.setText(tag);
        viewHolder.tag.setVisibility(tag.equals("") ? View.GONE : View.VISIBLE);
        viewHolder.description.setText(modInfoObject.getSubtitle());
        viewHolder.restore.setVisibility(modInfoObject.getModInfo().getMod().getOldFiles().isEmpty() ? View.GONE : View.VISIBLE);
        viewHolder.restore.setOnClickListener(v -> {
            ModRollbackDialog dialog = new ModRollbackDialog(getContext(), new ArrayList<>(modInfoObject.getModInfo().getMod().getOldFiles()), localModFile -> {
                modListPage.rollback(modInfoObject.getModInfo(), localModFile);
                notifyDataSetChanged();
            });
            dialog.show();
        });
        viewHolder.info.setOnClickListener(v -> {
            ModInfoDialog dialog = new ModInfoDialog(getContext(), modInfoObject);
            dialog.show();
        });
        Task.supplyAsync(() -> {
            for (RemoteMod.Type type : RemoteMod.Type.values()) {
                try {
                    if (modInfoObject.getRemoteMod() == null) {
                        Optional<RemoteMod.Version> remoteVersion = type.getRemoteModRepository().getRemoteVersionByLocalFile(modInfoObject.getModInfo(), modInfoObject.getModInfo().getFile());
                        if (remoteVersion.isPresent()) {
                            RemoteMod remoteMod = type.getRemoteModRepository().getModById(remoteVersion.get().getModid());
                            modInfoObject.getModInfo().setRemoteVersion(remoteVersion.get());
                            modInfoObject.setRemoteMod(remoteMod);
                        } else {
                            continue;
                        }
                    }
                    return modInfoObject.getRemoteMod();
                } catch (Throwable ignore) {
                }
            }
            return null;
        }).whenComplete(Schedulers.androidUIThread(), (remoteMod, exception) -> {
            if ((int) viewHolder.icon.getTag() == i && remoteMod != null) {
                viewHolder.icon.setVisibility(View.VISIBLE);
                Glide.with(viewHolder.icon).load(remoteMod.getIconUrl()).into(viewHolder.icon);
                viewHolder.name.setText(remoteMod.getTitle());
            }
        }).start();
        return view;
    }

    private String getTag(ModListPage.ModInfoObject modInfoObject) {
        StringBuilder stringBuilder = new StringBuilder();
        String modLoaderType = getModLoader(modInfoObject.getModInfo().getModLoaderType());
        stringBuilder.append(modLoaderType);
        if (modInfoObject.getMod() != null && LocaleUtils.isChinese(getContext())) {
            String pre = modLoaderType.equals("") ? "" : "   ";
            stringBuilder.append(pre).append(modInfoObject.getMod().getDisplayName());
        }
        return stringBuilder.toString();
    }

    private String getModLoader(ModLoaderType modLoaderType) {
        switch (modLoaderType) {
            case FORGE:
                return getContext().getString(R.string.install_installer_forge);
            case NEO_FORGED:
                return getContext().getString(R.string.install_installer_neoforge);
            case FABRIC:
                return getContext().getString(R.string.install_installer_fabric);
            case LITE_LOADER:
                return getContext().getString(R.string.install_installer_liteloader);
            case QUILT:
                return getContext().getString(R.string.install_installer_quilt);
            default:
                return "";
        }
    }
}

File: FCL/src/main/java/com/tungsten/fcl/ui/manage/LocalModListAdapter.kt
package com.tungsten.fcl.ui.manage

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.content.res.AppCompatResources
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.mio.ui.adapter.ViewHolder
import com.tungsten.fcl.R
import com.tungsten.fcl.activity.MainActivity
import com.tungsten.fcl.databinding.ItemLocalModBinding
import com.tungsten.fcl.ui.download.DownloadUI
import com.tungsten.fcl.ui.manage.ModListPage.ModInfoObject
import com.tungsten.fclcore.fakefx.beans.Observable
import com.tungsten.fclcore.fakefx.beans.property.ListProperty
import com.tungsten.fclcore.fakefx.beans.property.SimpleListProperty
import com.tungsten.fclcore.fakefx.collections.FXCollections
import com.tungsten.fclcore.fakefx.collections.ListChangeListener
import com.tungsten.fclcore.mod.LocalModFile
import com.tungsten.fclcore.mod.ModLoaderType
import com.tungsten.fclcore.mod.RemoteMod
import com.tungsten.fclcore.util.Logging
import com.tungsten.fclcore.util.StringUtils
import com.tungsten.fcllibrary.component.theme.ThemeEngine
import com.tungsten.fcllibrary.util.LocaleUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.Optional
import java.util.logging.Level
import kotlin.time.Duration.Companion.milliseconds

class LocalModListAdapter(
    private val context: Context,
    private val modListPage: ModListPage,
    val onChecked: () -> Unit
) :
    RecyclerView.Adapter<ViewHolder>() {
    val listProperty: ListProperty<ModInfoObject> = SimpleListProperty(
        FXCollections.observableArrayList()
    )
    val selectedItemsProperty: ListProperty<ModInfoObject?> =
        SimpleListProperty(
            FXCollections.observableArrayList<ModInfoObject?>()
        )

    val drawable = AppCompatResources.getDrawable(context, R.drawable.ic_cube)!!
    private val jobs = HashMap<String, Job>()

    fun listProperty(): ListProperty<ModInfoObject> {
        return listProperty
    }

    fun selectedItemsProperty(): ListProperty<ModInfoObject?> {
        return selectedItemsProperty
    }

    fun selectAll() {
        selectedItemsProperty.clear()
        selectedItemsProperty.addAll(listProperty)
    }

    fun selectInvert() {
        val list = ArrayList<ModInfoObject>()
        listProperty.forEach {
            if (!selectedItemsProperty.get().contains(it)) {
                list.add(it)
            }
        }
        selectedItemsProperty.clear()
        selectedItemsProperty.addAll(list)
    }

    private var fromSelf = false

    init {
        this.listProperty.addListener(ListChangeListener { c -> // 增量插入只通知新条目（notifyItemRangeInserted），已显示条目不会重绘，
            // 避免其在每次增量刷新时被重绘（图标重新加载、状态闪烁）
            var replaced = false
            while (c.next()) {
                if (c.wasReplaced()) {
                    // 全量替换（搜索 / 勾选筛选）：整体刷新，远程查询与选中状态全部重置
                    replaced = true
                } else if (c.wasRemoved()) {
                    c.removed.forEach {
                        jobs.remove(it.modInfo.fileName)?.cancel()
                        fromSelf = true
                        selectedItemsProperty.remove(it)
                        fromSelf = false
                    }
                    notifyItemRangeRemoved(c.from, c.removedSize)
                } else if (c.wasAdded()) {
                    notifyItemRangeInserted(c.from, c.addedSize)
                }
            }
            if (replaced) {
                jobs.values.forEach { it.cancel() }
                jobs.clear()
                fromSelf = true
                selectedItemsProperty.clear()
                fromSelf = false
                notifyDataSetChanged()
            }
        })
        selectedItemsProperty.addListener { _: Observable? ->
            if (!fromSelf) {
                notifyDataSetChanged()
            }
        }
    }

    private fun getTag(modInfoObject: ModInfoObject): String {
        val stringBuilder = StringBuilder()
        val modLoaderType = getModLoader(modInfoObject.modInfo.modLoaderType)
        stringBuilder.append(modLoaderType)
        if (modInfoObject.mod != null && LocaleUtils.isChinese(context)) {
            val pre = if (modLoaderType == "") "" else "   "
            stringBuilder.append(pre).append(modInfoObject.mod.getDisplayName())
        }
        return stringBuilder.toString()
    }

    private fun getModLoader(modLoaderType: ModLoaderType): String {
        return when (modLoaderType) {
            ModLoaderType.FORGE -> context.getString(R.string.install_installer_forge)
            ModLoaderType.NEO_FORGED -> context.getString(R.string.install_installer_neoforge)
            ModLoaderType.FABRIC -> context.getString(R.string.install_installer_fabric)
            ModLoaderType.LITE_LOADER -> context.getString(R.string.install_installer_liteloader)
            ModLoaderType.QUILT -> context.getString(R.string.install_installer_quilt)
            else -> ""
        }
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        return ViewHolder(
            ItemLocalModBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ).root
        )
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) {
        val binding = ItemLocalModBinding.bind(holder.itemView)
        val modInfoObject = listProperty[position]
        val key = modInfoObject.modInfo.fileName
        jobs[key]?.cancel()
        jobs.remove(key)
        binding.parent.backgroundTintList = ColorStateList(
            arrayOf<IntArray?>(intArrayOf()),
            intArrayOf(
                if (selectedItemsProperty.contains(modInfoObject)) ThemeEngine.getInstance()
                    .getTheme().getColor() else ThemeEngine.getInstance().getTheme()
                    .ltColor
            )
        )
        ThemeEngine.getInstance().unregisterEvent(binding.root)
        ThemeEngine.getInstance().registerEvent(binding.root) {
            binding.parent.backgroundTintList = ColorStateList(
                arrayOf<IntArray?>(intArrayOf()),
                intArrayOf(
                    if (selectedItemsProperty.contains(modInfoObject)) ThemeEngine.getInstance()
                        .getTheme().color else ThemeEngine.getInstance().getTheme()
                        .ltColor
                )
            )
        }
        binding.parent.setOnClickListener { _: View? ->
            if (selectedItemsProperty.contains(modInfoObject)) {
                fromSelf = true
                selectedItemsProperty.remove(modInfoObject)
                fromSelf = false
                binding.parent.backgroundTintList = ColorStateList(
                    arrayOf<IntArray?>(
                        intArrayOf()
                    ), intArrayOf(ThemeEngine.getInstance().getTheme().ltColor)
                )
            } else {
                fromSelf = true
                selectedItemsProperty.add(modInfoObject)
                fromSelf = false
                binding.parent.backgroundTintList = ColorStateList(
                    arrayOf<IntArray?>(
                        intArrayOf()
                    ), intArrayOf(ThemeEngine.getInstance().getTheme().getColor())
                )
            }
        }
        //必须先清除Listener
        binding.check.setOnCheckedChangeListener(null)
        binding.check.isChecked = modInfoObject.active.get()
        binding.check.setOnCheckedChangeListener { _, checked ->
            modInfoObject.active.set(checked)
            onChecked.invoke()
        }
        binding.name.text = modInfoObject.title
        val tag = getTag(modInfoObject)
        binding.tag.text = tag
        binding.tag.visibility = if (tag == "") View.GONE else View.VISIBLE
        binding.description.text = modInfoObject.subtitle
        binding.restore.visibility = if (modInfoObject.modInfo.mod.oldFiles
                .isEmpty()
        ) View.GONE else View.VISIBLE
        binding.restore.setOnClickListener {
            val dialog = ModRollbackDialog(
                context,
                ArrayList<LocalModFile?>(modInfoObject.modInfo.mod.oldFiles)
            ) { localModFile: LocalModFile? ->
                modListPage.rollback(modInfoObject.modInfo, localModFile)
                notifyDataSetChanged()
            }
            dialog.show()
        }
        binding.info.setOnClickListener {
            val dialog = ModInfoDialog(context, modInfoObject)
            dialog.show()
        }
        binding.jump.visibility = View.GONE
        binding.jump.setOnClickListener {
            val uiManager = MainActivity.getInstance().uiManager
            MainActivity.getInstance().binding.download.isSelected = true
            uiManager.downloadUI.showDownloadPage(DownloadUI.PAGE_ID_DOWNLOAD_MOD)
            uiManager.downloadUI.downloadPage.jumpToModPage(modInfoObject.remoteMod)
        }

        drawable.setTint(ThemeEngine.getInstance().getTheme().getColor())
        binding.icon.setImageDrawable(drawable)
        val cachedRemoteMod = modInfoObject.remoteMod
        if (cachedRemoteMod != null) {
            applyRemoteMod(binding, cachedRemoteMod, modInfoObject)
        }
    }

    /**
     * 远程信息查询挂在 attach 生命周期上而不是 onBindViewHolder：RecyclerView 的 view cache
     * （视口外侧各约 2 个条目）复用缓存 view 重新显示时不会重新绑定，挂在绑定上会导致
     * 查询被防抖跳过后（如快速滑动）永久失去重试机会；attach 则每次重新显示都会触发。
     * detach 时取消在途查询，delay(200ms) 后仍存活即说明条目停在了屏幕上。
     */
    override fun onViewAttachedToWindow(holder: ViewHolder) {
        super.onViewAttachedToWindow(holder)
        val position = holder.bindingAdapterPosition
        if (position == RecyclerView.NO_POSITION || position >= listProperty.size) return
        val modInfoObject = listProperty[position]
        if (modInfoObject.remoteMod != null) return
        val key = modInfoObject.modInfo.fileName
        val existing = jobs[key]
        if (existing != null && existing.isActive) return
        val binding = ItemLocalModBinding.bind(holder.itemView)
        holder.itemView.tag = key
        val job = MainActivity.getInstance().lifecycleScope.launch {
            delay(200L.milliseconds)
            val mod = withContext(Dispatchers.IO) {
                for (type in RemoteMod.Type.entries.toTypedArray()) {
                    ensureActive()
                    try {
                        if (modInfoObject.remoteMod == null) {
                            val remoteVersion: Optional<RemoteMod.Version?> =
                                type.remoteModRepository.getRemoteVersionByLocalFile(
                                    modInfoObject.modInfo,
                                    modInfoObject.modInfo.file
                                )
                            if (remoteVersion.isPresent) {
                                val remoteMod: RemoteMod? = type.remoteModRepository
                                    .getModById(remoteVersion.get().modid())
                                modInfoObject.modInfo.remoteVersion =
                                    remoteVersion.get()
                                modInfoObject.remoteMod = remoteMod
                            } else {
                                continue
                            }
                        }
                        return@withContext modInfoObject.remoteMod
                    } catch (e: Throwable) {
                        Logging.LOG.log(
                            Level.SEVERE,
                            "getRemoteVersionByLocalFile error: ${modInfoObject.modInfo.file.fileName}\n${e.toString()}"
                        )
                    }
                }
                null
            }
            if (isActive) {
                mod?.let {
                    applyRemoteMod(binding, it, modInfoObject)
                }
            }
        }
        jobs[key] = job
    }

    override fun onViewDetachedFromWindow(holder: ViewHolder) {
        super.onViewDetachedFromWindow(holder)
        val key = holder.itemView.tag as? String ?: return
        holder.itemView.tag = null
        jobs[key]?.cancel()
        jobs.remove(key)
    }

    @SuppressLint("SetTextI18n")
    private fun applyRemoteMod(
        binding: ItemLocalModBinding,
        mod: RemoteMod,
        modInfoObject: ModInfoObject
    ) {
        binding.icon.visibility = View.VISIBLE
        Glide.with(binding.icon).load(mod.iconUrl).error(drawable)
            .into(binding.icon)
        binding.name.text = mod.title
        binding.jump.visibility = View.VISIBLE
        if (modInfoObject.mod != null && LocaleUtils.isChinese(context)) {
            val name = modInfoObject.mod.name()
            if (name.isNotEmpty() && StringUtils.containsChinese(name)) {
                binding.name.text = "[${name}]${mod.title}"
            }
        }
    }

    override fun getItemCount(): Int {
        return listProperty.size
    }
}
